New features: override version, detect git URLs, ...

iarspider opened the pull request:
specify Git branch/tag/commit on the command line

iarspider committed:
New features: override version, detect git URLs, specify branch/tag/commit on the command line

iarspider committed:
New features: override version, detect git URLs, specify branch/tag/commit on the command line

iarspider commented:
It would be nice to automatically get list of branches/tags in the repo, but my spack-foo is too weak to figure out how to do it...

iarspider committed:
Flake-8

adamjstewart commented on lib/spack/spack/cmd/create.py:
```diff
@@ -730,6 +746,20 @@ def get_url(args):
+        bool: True if it seems to be a git repository
+    """
+
+    return url.startswith('git://') or url.startswith('git@') or url.endswith('.git')
```
What about https://github.com/foo/bar? If there are no other things tacked onto the end it would be valid. May need a regex to test this.

iarspider replied:
As far as I understand, in general `.git` is required - Github forgives missing `.git` quietly, Gitlab does a redirect (and git warns about it). Some git servers could be less forgiving and just return an error.

adamjstewart replied:
For our current packages, it looks like 22 out of 171 packages that use the `git` attribute don't append `.git`. We might want to make a special exception for GitHub.

iarspider replied:
> For our current packages, it looks like 22 out of 171 packages that use the `git` attribute don't append `.git`. We might want to make a special exception for GitHub.

... or fix those 22 package. Or even leave them as-is, since the PR only changes `create` command, it doesn't affect any existing packages.

iarspider replied:
@adamjstewart  After doing a few checks with different git servers (github, gitlab, whatever linux kernel git uses), I find that the `.git` suffix is indeed optional. I have now added `--git` switch to force using the `url` argument passed to `spack create` as a value for `git` attribute rather than `url` one. Let me know if you think removing the detection code and relying on that switch will be a better solution (and implying `--git` if `--branch`, `--tag` or `--commit` are set)

iarspider committed:
Update completion

iarspider committed:
Add `--git` flag

* Update create.py

* Update spack-completion.bash

iarspider committed:
Update packaging_guide.rst

iarspider committed:
Merge branch 'develop' into feature/spack-create-git

iarspider committed:
Flake-8

iarspider committed:
Flake-8

iarspider committed:
Fix completion

iarspider commented:
@adamjstewart gentle ping

adamjstewart commented on lib/spack/spack/cmd/create.py:
```diff
@@ -730,6 +749,30 @@ def get_url(args):
+
+
+def force_git_url(url):
+    """ Dummy function to replace is_git_url if --git argument is used"""
```
Needs args/returns in docstring. Also, is this function really necessary?

iarspider replied:
Rewrote the code and removed is_git_url/force_git_url functions

iarspider replied:
Actually had to restore them. I need the `is_git_url` check in `get_versions`, and I didn't want to change the signature of that function to somehow pass `--git` flag into it.

adamjstewart commented on lib/spack/spack/cmd/create.py:
```diff
@@ -755,10 +798,37 @@ def get_versions(args, name):
     # version('1.2.4')"""
 
+    # Default git-based version
+    git_versions = """
```
Is this default ever used?

iarspider replied:
Fixed

adamjstewart commented on lib/spack/spack/cmd/create.py:
```diff
@@ -755,10 +798,37 @@ def get_versions(args, name):
+            if args.version is not None:
+                _version = _version.format(args.version, 'commit', args.commit)
+            else:
+                _version = '    #FIXME: add proper version\n' + \
```
```suggestion
                _version = '    # FIXME: add proper version\n' + \
```

iarspider committed:
Update lib/spack/spack/cmd/create.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

iarspider committed:
Changes from reviee

iarspider committed:
Changes from reviee

iarspider committed:
Update create.py

iarspider committed:
Revert "Changes from reviee"

This reverts commit 08fa2efa5a100144860f327bead4012b1c876ecc.

iarspider committed:
Update create.py

iarspider committed:
Update create.py

tldahlgren commented:
@adamjstewart ping  Someone was asking about this one being added to v0.18

adamjstewart commented:
@tldahlgren I'll let someone else handle this one, I'm pretty much only maintaining python/geospatial packages these days, don't have time to maintain core.

tldahlgren commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -57,10 +57,14 @@ Here's an example:
+
+.. note::
+
+    You can override version detected by Spack with `--version` argument
```
```suggestion
    You can override the version detected by Spack using the `--version` argument.
```

tldahlgren commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -1078,6 +1082,16 @@ Submodules
 
+.. note::
+
+    Spack will detect git-like url when you run `spack create` by checking
```
```suggestion
    Spack will detect a git-like url when you run `spack create` by checking
```

tldahlgren commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -1078,6 +1082,16 @@ Submodules
+.. note::
+
+    Spack will detect git-like url when you run `spack create` by checking
+    for `git@` prefix or `.git` suffix. One can also use `--git` flag to signal
```
```suggestion
    for `git@` prefix or `.git` suffix. Or you can use `--git` flag to signal
```

tldahlgren commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -1078,6 +1082,16 @@ Submodules
+    Spack that the URL passed to `spack create` is not a direct URL, but a Git
+    repository URL.
+    
+    You can also specify tag (with `--tag` argument), commit (`--commit`) or branch (`--branch`). By default,
```
```suggestion
    You can also specify a tag (with `--tag` argument), a commit (`--commit`) or a branch (`--branch`). By default,
```

tldahlgren commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -1078,6 +1082,16 @@ Submodules
+    repository URL.
+    
+    You can also specify tag (with `--tag` argument), commit (`--commit`) or branch (`--branch`). By default,
+    Spack will use branch or tag name, or commit id as version, you can override it with `--version` argument.
```
```suggestion
    Spack will use the branch, tag name, or commit id as the version, which you can override with the `--version` argument.
```

tldahlgren commented on lib/spack/spack/cmd/create.py:
```diff
@@ -578,6 +582,21 @@ def setup_parser(subparser):
+        help="use git to download source from repository passed in url argument")
+    subparser.add_argument(
+        '-V', '--version',
+        help='Force package version')
```
```suggestion
        help='override derived package version')
```

tldahlgren commented on lib/spack/spack/cmd/create.py:
```diff
@@ -578,6 +582,21 @@ def setup_parser(subparser):
+        help='Force package version')
+    group = subparser.add_mutually_exclusive_group()
+    group.add_argument('-B', '--branch',
+                       help='specify branch of git repository. Not recommended,'
```
Not sure `Not recommended` applies at least in the case of `develop`, `main`, and `master`.

tldahlgren commented on lib/spack/spack/cmd/create.py:
```diff
@@ -578,6 +582,21 @@ def setup_parser(subparser):
+    group.add_argument('-B', '--branch',
+                       help='specify branch of git repository. Not recommended,'
+                       ' use `--commit` instead. Only used for git URLs')
+    group.add_argument('-T', '--tag',
+                       help='specify tag of git repository. Not recommended,'
+                       ' use `--commit` instead. Only used for git URLs')
+    group.add_argument('-C', '--commit',
+                       help='specify commit of git repository. Only used for git URLs')
```
For some reason I am under the impression `tag`s are "better than" `commit`s.

iarspider replied:
Commits will never change, tags can.

tldahlgren replied:
I'll try to remember that.  Thanks.

tldahlgren commented on lib/spack/spack/cmd/create.py:
```diff
@@ -730,6 +749,38 @@ def get_url(args):
     return url
 
 
+def is_git_url(url):
+    """Check if the URL is likely to be a git repository. The code doesn't attempt
+    to clone the repository!
+
+    Args:
+        url (str): The url to check
+
+    Returns:
+        bool: True if it seems to be a git repository
+    """
+
+    try:
+        spack.util.url.parse_git_url(url)
+    except ValueError:
+        return False
```
It is not sufficient to check the exception.  For example, a badly formatted string, such as`https:/this.does.not.exist`, will not raise `ValueError`. 

Perhaps at least check that the first element -- the scheme (e.g., `git`, `ssh`, `http`, or `https`) -- is not `None`.

It is also worth noting that this will return `True` even if the `url` is not for a repository (e.g., `https://github.com/spack/spack/blob/develop/CHANGELOG.md` and `https://github.com/spack/spack/archive/refs/heads/develop.zip`).

iarspider replied:
Scheme can be empty (`git@github.com:spack/spack.git`), but I agree that maybe this logic needs improvement

tldahlgren replied:
In all (11) builtin packages using that form, the scheme is `ssh` (i.e., the URL is `ssh://git@...`). So one option would be to automatically prepend the scheme when the URL starts `git@`.  Alternatively, you can require a proper URL for it by requiring the scheme.

tldahlgren commented on lib/spack/spack/cmd/create.py:
```diff
@@ -730,6 +749,38 @@ def get_url(args):
+        return True
+
+
+def force_git_url(url):
+    """Dummy replacement for `is_git_url` to always return True.
+
+    Args:
+        url (str): The url to check
+
+    Returns:
+        bool: always True
+    """
+
+    return True
```
This is irrelevant given the current (May 18th) implementation of `is_git_url()`, which will return `True` even for an archive file.

tldahlgren commented on lib/spack/spack/cmd/create.py:
```diff
@@ -869,6 +958,12 @@ def get_repository(args, name):
 def create(parser, args):
+    global is_git_url
+
+    # Handle `--git` argument
+    if args.git:
+        is_git_url = force_git_url
```
This code is not necessary based on how `is_git_url` is currently written.

tldahlgren commented on lib/spack/spack/cmd/create.py:
```diff
@@ -755,10 +806,37 @@ def get_versions(args, name):
+                    _version.format(args.commit, 'commit', args.commit)
+        if args.tag is not None:
+            _version = _version.format(args.version or args.tag, 'tag', args.tag)
+        if args.branch is not None:
```
Can the option be empty?

tldahlgren commented on lib/spack/spack/cmd/create.py:
```diff
@@ -755,10 +806,37 @@ def get_versions(args, name):
+            else:
+                _version = '    # FIXME: add proper version\n' + \
+                    _version.format(args.commit, 'commit', args.commit)
+        if args.tag is not None:
```
Can this be empty?

tldahlgren commented on lib/spack/spack/cmd/create.py:
```diff
@@ -755,10 +806,37 @@ def get_versions(args, name):
+
+    if is_git_url(args.url) and has_git_option:
+        _version = "    version('{0}', {1}='{2}')"
+        if args.commit is not None:
```
Can this be empty?

tldahlgren commented on lib/spack/spack/cmd/create.py:
```diff
@@ -125,11 +125,15 @@ def install(self, spec, prefix):
         super(PackageTemplate, self).__init__(name, versions)
 
-        self.url_def = self.url_line.format(url=url)
+        if is_git_url(url):
+            self.url_def = self.git_line.format(url=url)
+        else:
+            self.url_def = self.url_line.format(url=url)
```
As written, `url_line` will not get used.

tldahlgren reviewed (commented):
IMO Support for these options could be simplified.

We're already generating a template for the package with, pre-PR, information derived from the archive file.  We also support packages with archive `url`s *and* `git`.

So why not add lines and guiding comments for the `git` assignment, etcetera?

Since the URLs for less than 80 of the ~1071 packages using `git` don't actually end in `.git`, you could use `.git` to determine whether to populate the `git` or `url` property and have a comment telling the user to correct the URL as needed.

You could also use that to determine if the branch, commit, or tag options are valid IF you want.

If it is determined that the `git` property will be populated, you could provide default `version('main', branch='main')` and `version('develop', branch='develop')` line(s) with a comment suggesting updates and or removals as needed.  This could also serve as an example that *might* discourage things like `version('develop', branch='master')`.

There are packages that have version directives covering two or more of `branch`, `commit`, and `tag` so IMO there's no harm in allowing the user to specify more than one with a line generated for each.  Examples for `branch`+{`commit`|`tag) include `amdlibm`, `ascent`, and `axom`.

iarspider committed:
Apply suggestions from code review (1/?)

Co-authored-by: Tamara Dahlgren <35777542+tldahlgren@users.noreply.github.com>

iarspider committed:
Merge branch 'spack:develop' into feature/spack-create-git

dev-zero committed:
cp2k: fix unbound var use without cuda (#30736)

fixes #30631

Jordan474 committed:
Don't try to mkdir upstream directory when nonexistent (#30744)

When an upstream is specified but the directory does not exist, don't
create the directory for it, it might not be yours.

haampie committed:
tar: add compress deps (#30641)

* remove spack dependency from package

* tar: fix compression programs, use pigz by default instead of gzip on -z

vvolkl committed:
cuda: use stage dir instead of /tmp during install (#29584)

Jordan474 committed:
Fix spack install chgrp on symlinks (#30743)

Fixes missing chgrp on symlinks in package installations, and errors on
symlinks referencing non-existent or non-writable locations.

Note: `os.chown(.., follow_symlinks=False)` is python3 only, but
`os.lchown` exists in both versions.

RemiLacroix-IDRIS committed:
libtheora: fix the hash of a patch. (#30740)

Cosmetic changes only, probably because gitlab.xiph.org got updated.

iarspider commented:
I have a new implementation in the works, will try to push tomorrow. 

@tldahlgren I think it is better to not detect git repository sometimes rather than wrongly detect regular URL as git, right?

iarspider committed:
Update millepede recipe (#30737)

kwryankrattiger committed:
Ascent: Add variant to disable blt_find_mpi (#30735)

This is needed to find MPI correctly on cray systems and similar.

ma595 committed:
fenicsx: update to v0.4 (#30661)

* Fix for xtensor-xsimd

* Add sha256 for all new releases

* renamed ufcx package

* Update sha for ffcx

* fixed hashes and modified fenics-dolfinx to depend on ufcx

* cleaned and fixed dependency types

* use spec.satisfies in cmake_args

* bumped to ufcx@0.4.1

* address PR comments

* fix hashes

* update parmetis in cmake_args to reflect default setting

* update versions

* renamed ufcx package

* fixed hashes and modified fenics-dolfinx to depend on ufcx

* cleaned and fixed dependency types

* use spec.satisfies in cmake_args

* bumped to ufcx@0.4.1

* address PR comments

* fix hashes

* update parmetis in cmake_args to reflect default setting

* update versions

* Add dependency fix

* bump basix to 0.4.2 and address PR comments

* Versioning fixes

* Use xtensor-0.24: and loosen pybind11

* Add conflicts for partitioners

* Updates on partitioners

* use define_from_variant

* Tidy up some dependencies

* Work on multi-variants for graph partitioners

* Fix KaHIP issue.

KaHIP changed the name of its library from 'interface' to 'kahip'. Pin earlier versions of DOLFINx to earlier verisons of KaHIP for proper detection.

Co-authored-by: Chris Richardson <chris@bpi.cam.ac.uk>
Co-authored-by: Garth N. Wells <gnw20@cam.ac.uk>

haampie committed:
python: more +optimizations (#30742)

Jordan474 committed:
Non-existent upstream is not fatal (#30746)

A non-existent upstream should not be fatal: it could only mean it is
not deployed yet. In the meantime, it should not block the user to
rebuild anything it needs.

A warning is still emitted, to let the user decide if this is ok or not.

wyphan committed:
Run scheduled CI workflows only in the main repo (#30729)

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

iarspider committed:
Update py-onnx-runtime to 1.10; update CMS patch (#30725)

* Update py-onnx-runtime to 1.10; update CMS patch

* Update package.py

manuelakuhn committed:
py-fracridge: add new package (#30739)

scheibelp committed:
Petsc: fix enable-x for virtuals (#30749)

* If direct dependencies provide virtuals, add those virtual names as well
* Also refer to package by virtual name for jpeg

becker33 committed:
errors: model error messages as an optimization problem (#30669)

Error messages for the clingo concretizer have proven challenging. The current messages are incredibly vague and often don't help users at all. Unsat cores in clingo are not guaranteed to be minimal, and lead to cores that are either not useful or need to be post-processed for hours to reach a minimal core.

Following up on an idea from a slack conversation with kwryankrattiger on slack, this PR takes a new approach. We eliminate most integrity constraints and minima/maxima on choice rules in clingo, and instead force invalid states to imply an error predicate. The error predicate can include context on the cause of the error (Package, Version, etc). These error predicates are then heavily optimized against, to ensure that we do not include error facts in the solution when a solution with no error facts could be generated. When post-processing the clingo solution to construct specs, any error facts cause the program to raise an exception. This leads to much more legible error messages. Each error predicate includes a priority and an error message. The error message is formatted by the remaining arguments to produce the error message. The priority is used to ensure that when clingo has a choice of which rules to violate, it chooses the one which will be most informative to the user.

Performance:

"fresh" concretizations appear to suffer a ~20% performance penalty under this branch, while "reuse" concretizations see a speedup of around 33%. 

Possible optimizations if users still see unhelpful messages:

There are currently 3 levels of priority of the error messages. Additional priorities are possible, and can allow us finer granularity to ensure more informative error messages are provided in lieu of less informative ones.

Future work:

Improve tests to ensure that every possible rule implying an error message is exercised

Jordan474 committed:
qt: Qt 5.15.0 requires OpenSSL 1.1.1 (#30754)

Fixes qt configure errors with external openssl on older systems (rhel7)

See
https://github.com/qt/qtbase/blob/efc02f9cc301f98c77079adae026ffd07f50d5ab/dist/changes-5.15.0#L346

This means for now on, `qt ^openssl@1.0` gets you `qt@5.15.4 ~ssl`:
clingo chooses latest qt version **but disables ssl support**.

ndevelder committed:
Add newer openfast versions and preliminary OpenMP compile support (#30752)

* Add version 3.0 and 3.1 and prelim OpenMP support

* Fix flag handler missing spec variable

* Use self.compiler.openmp_flag instead of -fopenmp

* Fix whitespace

iarspider committed:
DWZ: use virtual "elf" package (#30761)

manuelakuhn committed:
py-pybids: add 0.8.0 and 0.15.1 (#30756)

manuelakuhn committed:
py-formatizer: add new package (#30755)

qwertos committed:
[py-openslide-python] added verion 1.1.2 (#30722)

* [py-openslide-python] added verion 1.1.2 and set max py-setuptools version for 1.1.1

* [py-openslide-python]

- setuptools required for all possible newer versions
- python is type build run

* [py-openslide-python] use pil provider

srekolam committed:
update checksum and url for mlirmiopen recipe (#30771)

Commit:
visit: Overhaul to build in the DAV SDK (#30594)

* mesa-glu and mesa-demos: Fix conflicts with glu and osmesa

* visit: Update visit dependencies

* ecp-data-vis-sdk: Enable +visit

* ci[data-vis-sdk]: Enable +visit

glennpj committed:
Fix toolchain detection for oneapi/dpcpp compilers (#30775)

The oneapi and dpcpp compilers are essentially the same except for which
binary is used foc CXX. Spack will detect them as "mixed toolchain" and
not inject compiler optimization flags. This will be needed once
archspec has entries for the oneapi and dpcpp compilers. This PR detects
when dpcpp and oneapi are in the toolchains list and explicitly sets
`is_mixed_toolchain` to `False`.

dependabot[bot] committed:
build(deps): bump actions/upload-artifact from 3 to 3.1.0 (#30778)

Bumps [actions/upload-artifact](https://github.com/actions/upload-artifact) from 3 to 3.1.0.
- [Release notes](https://github.com/actions/upload-artifact/releases)
- [Commits](https://github.com/actions/upload-artifact/compare/v3...3cea5372237819ed00197afe530f5a7ea3e805c8)

---
updated-dependencies:
- dependency-name: actions/upload-artifact
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

snehring committed:
paml: add v4.10.3 (#30772)

manuelakuhn committed:
py-ww: add new package (#30767)

* py-ww: add new package

* Add missing py-pytest-runner dependency

qwertos committed:
New package: py-motor (#30763)

* New package: py-motor

* Fixed dependency errors

* Fixed flake style errors

* Fixed linked issue

Co-authored-by: Viv Eric Hafener <vehrc@sporcbuild.rc.rit.edu>

haampie committed:
gcc: add `build_type` and `profiled` variants (#30660)

Add a `build_type` variant, which allows building optimized compilers,
as well as target libraries (libstdc++ and friends).

The default is `build_type=RelWithDebInfo`, which corresponds to GCC's
default of -O2 -g.

When building with `+bootstrap %gcc`, also add Spack's arch specific
flags using the common denominator between host and new GCC.

It is done by creating a config/spack.mk file in def patch, that looks
as follows:
```
BOOT_CFLAGS := $(filter-out -O% -g%, $(BOOT_CFLAGS)) -O2 -g -march=znver2 -mtune=znver2
CFLAGS_FOR_TARGET := $(filter-out -O% -g%, $(CFLAGS_FOR_TARGET)) -O2 -g -march=znver2 -mtune=znver2
CXXFLAGS_FOR_TARGET := $(filter-out -O% -g%, $(CXXFLAGS_FOR_TARGET)) -O2 -g -march=znver2 -mtune=znver2
```

asarkar-parsys committed:
Feature/composed boost pkg deps p3 (#28961)

* This commit removes the Boost.with_default_variants to variants that packages are precisely dependant upon. This is the third batch of 16 packages with modified boost dependencies.

* style fix

* Update var/spack/repos/builtin/packages/sympol/package.py

Co-authored-by: Tim Haines <thaines.astro@gmail.com>

* fix style

* Apply suggestions from code review

Co-authored-by: Tim Haines <thaines.astro@gmail.com>

* Fix Trilinos boost deps

* Fix style

Co-authored-by: Tim Haines <thaines.astro@gmail.com>
Co-authored-by: Tom Scogland <tom.scogland@gmail.com>

kuberry committed:
Add Teuchos patch that fixes implicit casting of complex numbers (#30777)

MichaelLaufer committed:
py-pyfr: Add v1.14.0, add LIBXSMM variant (#30612)

* Update PyFR, Gimmik versions. Add PyFR LIBXSMM variant

* Fixes

* Changes to variants

* Update py-gimmik version requirement in py-pyfr

* Revert "Update py-gimmik version requirement in py-pyfr"

This reverts commit 3b3fde30425ece252290eb0a2793ca1707f84860.

* Update libxsmm conflicts

tuxfan committed:
gdb: add v11.2 (#30780)

- This resolves bug https://sourceware.org/PR28302

alalazo committed:
llvm: make "omp_as_runtime" variant conditional (#30782)

fixes #30700

To avoid clingo adding penalties for not using the
default value for a variant, it's better to model
the variant as conditional where possible.

jeanbez committed:
hdf5-vol-async: update new version, tests, and runtime envs (#30713)

* Update h5bench maintainers and versions

* Include version 1.1 for h5bench

* Correct release hash and set default version

* Update .tar.gz version

* Include new version and update runtime

* Update year

* Update package.py

* Update package.py

manuelakuhn committed:
py-rnc2rng: fix 2.6.5 and add 2.6.6 (#30784)

balos1 committed:
xsdk-examples: add v0.3.0 (#30770)

Co-authored-by: Satish Balay <balay@mcs.anl.gov>
Co-authored-by: Veselin Dobrev <dobrev@llnl.gov>

marcus-elia committed:
abseil-cpp: add v20211102 (#30785)

Commit:
ci: Map visit to huge instance for the data-vis-sdk pipeline (#30779)

w8jcik committed:
plumed: add versions up to v2.8.0 (#30787)

haampie committed:
Deprecate `spack:concretization` over `concretizer:unify`  (#30038)

* Introduce concretizer:unify option to replace spack:concretization

* Deprecate concretization

* Make spack:concretization overrule concretize:unify for now

* Add environment update logic to move from spack:concretization to spack:concretizer:reuse

* Migrate spack:concretization to spack:concretize:unify in all locations

* For new environments make concretizer:unify explicit, so that defaults can be changed in 0.19

alalazo committed:
archspec: add oneapi and dpcpp flag support (#30783)

snehring committed:
genemark-et: updating to 4.69 (#30793)

snehring committed:
usearch: adding in new version, updating checksums (#30776)

bollig committed:
Added cloud_pipline for E4S on Amazon Linux (#29522)

Add two new cloud pipelines for E4S on Amazon Linux, include arm and x86 (v3 + v4) stacks.

Notes:
- Updated mpark-variant to remove conflict that no longer exists in Amazon Linux
- Which command on Amazon Linux prefixes on all results when padded_length is too high. In this case, padded_length<=503 works as expected. Chose conservative length of 384.

becker33 committed:
Documentation and new method for `CachedCMakePackage` build system (#22706)

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

qwertos committed:
[py-h2] added version 3.2.0 and 4.1.0 (#29804)

* [py-h2] py-wheel is implied by PythonPackage

* [py-h2] python dependencies should be type=('build', 'run')

* [py-h2] fixed dependencies for py-h2@4.0.0

* [py-h2] added version 3.2.0

* [py-h2] added version 4.1.0

* [py-h2] Older version requires py-enum34 for older versions of python

bvanessen committed:
Bugfix hwloc find cuda (#30788)

* Added autotools configure flags to ensure that hwloc finds the correct
version of CUDA that it was concretized against, rather than the first
one that package config finds.

* Added support for finding the correct version of ROCm libraries.  Fixed Flake8.

* Fixed guard on finding ROCm library

scottwittenburg committed:
Revert "Added cloud_pipline for E4S on Amazon Linux (#29522)" (#30796)

This reverts commit 07e9c0695af09bbf5d1ebe51bb6f2b34eb88a3df.

tgamblin committed:
specs: emit better parsing errors for specs. (#24860)

Parse error information is kept for specs, but it doesn't seem like we propagate it
to the user when we encounter an error.  This fixes that.

e.g., for this error in a package:

```python
    depends_on("python@:3.8", when="0.900:")
```

Before, with no context and no clue that it's even from a particular spec:

```
==> Error: Unexpected token: ':'
```

With this PR:

```
==> Error: Unexpected token: ':'
  Encountered when parsing spec:
    0.900:
         ^
```

trws committed:
strip -Werror: all specific or none (#30284)

Add a config option to strip `-Werror*` or `-Werror=*` from compile lines everywhere.

```yaml
config:
    keep_werror: false
```

By default, we strip all `-Werror` arguments out of compile lines, to avoid unwanted
failures when upgrading compilers.  You can re-enable `-Werror` in your builds if
you really want to, with either:

```yaml
config:
    keep_werror: all
```

or to keep *just* specific `-Werror=XXX` args:

```yaml
config:
    keep_werror: specific
```

This should make swapping in newer versions of compilers much smoother when
maintainers have decided to enable `-Werror` by default.

becker33 committed:
compiler flags: imposed hashes impose the lack of additional compiler flags (#30797)

adamjstewart committed:
gplates: add v2.3.0 (#30676)

prwolfe committed:
trilinos: add adelus, aprepro and teuchos variants (#28935)

Mr-Timn committed:
su2: add v7.3.1 (#30794)

balay committed:
petsc: update rocrand location wrt rocm@5.1.0 (#30790)

rocm-5.1.0 removed librocrand.so from ROCM_DIR/rocrand/lib location (but  includes are still at this location)

/opt/rocm-5.0.2/lib/librocrand.so
/opt/rocm-5.0.2/rocrand/lib/librocrand.so
/opt/rocm-5.1.0/lib/librocrand.so

drwxr-xr-x 2 root root 617 Mar  8 08:20 /opt/rocm-5.0.2/rocrand/include
drwxr-xr-x 2 root root 617 Mar 31 09:48 /opt/rocm-5.1.0/rocrand/include

bollig committed:
libfabric has needed rdma-core for efa since 1.10.0 (#30798)

iarspider committed:
vdt: add preload variant (#30030)

matz-e committed:
apptainer: new package (#30745)

edwardsp committed:
Update regex to correctly identify quoted args (#23494)

Previously the regex was only checking for presence of quotes as a beginning
or end character and not a matching set.  This erroneously identified the
following *single* argument as being quoted:

    source bashenvfile &> /dev/null && python3 -c "import os, json; print(json.dumps(dict(os.environ)))"

sethrj committed:
environments: fail gracefully on missing keys (#26378)

qwertos committed:
New package: py-x21 (#30225)

* Py-x21 now works, needs dependencies

Conflicts:
	var/spack/repos/rit-rc/packages/py-x21/package.py

* Added dependencies to py-x21

* Making flake style check happy

* [py-x21] flake8

* [py-x21]

- added homepage
- added placeholder description
- added comment about checksums

* [py-x21] added darwin support and fixed issue with python 3.7 wheel name

* [py-x21] adding checksum hash

* [py-x21] removed duplicate py-pynacl

* [py-x21]

- updated description
- updated version listing to have a different version for each version
  of python. Also, versions dependent on sys.platform
- updated url_for_version to not require post concretized information so
  that spack checksum works

* [py-x21] isort

Co-authored-by: vehrc <vehrc@rit.edu>

alalazo committed:
Best effort co-concretization (iterative algorithm) (#28941)

Currently, environments can either be concretized fully together or fully separately. This works well for users who create environments for interoperable software and can use `concretizer:unify:true`. It does not allow environments with conflicting software to be concretized for maximal interoperability.

The primary use-case for this is facilities providing system software. Facilities provide multiple MPI implementations, but all software built against a given MPI ought to be interoperable.

This PR adds a concretization option `concretizer:unify:when_possible`. When this option is used, Spack will concretize specs in the environment separately, but will optimize for minimal differences in overlapping packages.

* Add a level of indirection to root specs

This commit introduce the "literal" atom, which comes with
a few different "arities". The unary "literal" contains an
integer that id the ID of a spec literal. Other "literals"
contain information on the requests made by literal ID. For
instance zlib@1.2.11 generates the following facts:

literal(0,"root","zlib").
literal(0,"node","zlib").
literal(0,"node_version_satisfies","zlib","1.2.11").

This should help with solving large environments "together
where possible" since later literals can be now solved
together in batches.

* Add a mechanism to relax the number of literals being solved

* Modify spack solve to display the new criteria

Since the new criteria is above all the build criteria,
we need to modify the way we display the output.

Originally done by Greg in #27964 and cherry-picked
to this branch by the co-author of the commit.

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

* Inject reusable specs into the solve

Instead of coupling the PyclingoDriver() object with
spack.config, inject the concrete specs that can be
reused.

A method level function takes care of reading from
the store and the buildcache.

* spack solve: show output of multi-rounds

* add tests for best-effort coconcretization

* Enforce having at least a literal being solved

Co-authored-by: Greg Becker <becker33@llnl.gov>

snehring committed:
tbl2asn: adding currently available version (#30774)

skosukhin committed:
mpich: re-enable building of the older versions (#30766)

* mpich: enable building @3.4:3.4.3 ~cuda

* mpich: add dependency on mxm

masterleinad committed:
Allow Kokkos with OpenMPTarget backend (#30724)

* Allow Kokkos with OpenMPTarget backend

* Restrict SYCL and OpenMPTarget to C++17 or higher

* Improve C++ standard check for SYCL and OpenMPTarget

* Fix indentation

qwertos committed:
[libaec] Added version 1.0.6 (#30812)

Co-authored-by: James A Zilberman <jazrc@rit.edu>

alalazo committed:
Add a command to generate a local mirror for bootstrapping (#28556)

This PR builds on #28392 by adding a convenience command to create a local mirror that can be used to bootstrap Spack. This is to overcome the inconvenience in setting up this mirror manually, which has been reported when trying to setup Spack on air-gapped systems.

Using this PR the user can create a bootstrapping mirror, on a machine with internet access, by:

% spack bootstrap mirror --binary-packages /opt/bootstrap
==> Adding "clingo-bootstrap@spack+python %apple-clang target=x86_64" and dependencies to the mirror at /opt/bootstrap/local-mirror
==> Adding "gnupg@2.3: %apple-clang target=x86_64" and dependencies to the mirror at /opt/bootstrap/local-mirror
==> Adding "patchelf@0.13.1:0.13.99 %apple-clang target=x86_64" and dependencies to the mirror at /opt/bootstrap/local-mirror
==> Adding binary packages from "https://github.com/alalazo/spack-bootstrap-mirrors/releases/download/v0.1-rc.2/bootstrap-buildcache.tar.gz" to the mirror at /opt/bootstrap/local-mirror

To register the mirror on the platform where it's supposed to be used run the following command(s):
  % spack bootstrap add --trust local-sources /opt/bootstrap/metadata/sources
  % spack bootstrap add --trust local-binaries /opt/bootstrap/metadata/binaries
The mirror has to be moved over to the air-gapped system, and registered using the commands shown at prompt. The command has options to:

1. Add pre-built binaries downloaded from Github (default is not to add them)
2. Add development dependencies for Spack (currently the Python packages needed to use spack style)

* bootstrap: refactor bootstrap.yaml to move sources metadata out

* bootstrap: allow adding/removing custom bootstrapping sources

This operation can be performed from the command line since
new subcommands have been added to `spack bootstrap`

* Add --trust argument to spack bootstrap add

* Add a command to generate a local mirror for bootstrapping

* Add a unit test for mirror creation

qwertos committed:
[libdeflate] Added version 1.10 (#30813)

Co-authored-by: James A Zilberman <jazrc@rit.edu>

scottwittenburg committed:
buildcache: Update layout and signing (#30750)

This PR introduces a new build cache layout and package format, with improvements for
both efficiency and security.

## Old Format
Currently a binary package consists of a `spec.json` file at the root and a `.spack` file,
which is a `tar` archive containing a copy of the `spec.json` format, possibly a detached
signature (`.asc`) file, and a tar-gzip compressed archive containing the install tree.

```
build_cache/
  # metadata (for indexing)
  <arch>-<compiler>-<name>-<ver>-24zvipcqgg2wyjpvdq2ajy5jnm564hen.spec.json
  <arch>/
    <compiler>/
      <name>-<ver>/
        # tar archive
        <arch>-<compiler>-<name>-<ver>-24zvipcqgg2wyjpvdq2ajy5jnm564hen.spack
          # tar archive contents:
          # metadata (contains sha256 of internal .tar.gz)
          <arch>-<compiler>-<name>-<ver>-24zvipcqgg2wyjpvdq2ajy5jnm564hen.spec.json
          # signature
          <arch>-<compiler>-<name>-<ver>-24zvipcqgg2wyjpvdq2ajy5jnm564hen.spec.json.asc
          # tar.gz-compressed prefix
          <arch>-<compiler>-<name>-<ver>-24zvipcqgg2wyjpvdq2ajy5jnm564hen.tar.gz
```

After this change, the nesting has been removed so that the `.spack` file is the
compressed archive of the install tree.  Now signed binary packages, will take the
form of a clearsigned `spec.json` file (a `spec.json.sig`) at the root, while unsigned
binary packages will contain a `spec.json` at the root.

## New Format

```
build_cache/
  # metadata (for indexing, contains sha256 of .spack file)
  <arch>-<compiler>-<name>-<ver>-24zvipcqgg2wyjpvdq2ajy5jnm564hen.spec.json
  # clearsigned spec.json metadata
  <arch>-<compiler>-<name>-<ver>-24zvipcqgg2wyjpvdq2ajy5jnm564hen.spec.json.sig
  <arch>/
    <compiler>/
      <name>-<ver>/
        # tar.gz-compressed prefix (may support more compression formats later)
        <arch>-<compiler>-<name>-<ver>-24zvipcqgg2wyjpvdq2ajy5jnm564hen.spack
```

## Benefits
The major benefit of this change is that the signatures on binary packages can be
verified without:

1. Having to download the tarball, or
2. having to extract an unknown tarball.

(1) is an improvement in efficiency; (2) is a security fix: we now ensure that we trust the
binary before we try to run it through `tar`, which avoids potential attacks.

## Backward compatibility
Also after this change, spack should still be able to handle the previous buildcache
structure and binary mirrors with mixed layouts.

qwertos committed:
[libwebp] Added version 1.2.2 (#30814)

Co-authored-by: James A Zilberman <jazrc@rit.edu>

mdorier committed:
[leveldb] add patch to fix check for -Wthread-safety (#30810)

drkrynstrng committed:
Add rsync 3.2.4 (#30821)

drkrynstrng committed:
Add aria2 1.36.0 (#30822)

Commit:
silo: Fix HDF5 1.13 API breakage (#30786)

adamjstewart committed:
py-pillow-simd: mark conflicts with aarch64 (#30828)

qwertos committed:
[lcms] Added version 2.13.1 and URL version (#30811)

Co-authored-by: James A Zilberman <jazrc@rit.edu>

haampie committed:
mbedtls: add conflicts over inline asm trouble with clang@12: (#30801)

haampie committed:
julia: support clang, set llvm NDEBUG correctly (#30800)

eugeneswalker committed:
tau: add v2.31.1 (#30820)

haampie committed:
dsfmt: set CC=cc (#30842)

haampie committed:
p7zip: fix %clang (#30843)

eschnett committed:
reprimand: update homepage (#30840)

adamjstewart committed:
py-scipy: add v1.8.1 (#30831)

haampie committed:
ccache: add missing pkgconfig dep (#30846)

drbenmorgan committed:
geant4: new version v11.0.2 (#30847)

mdorier committed:
[mochi-margo] added version 0.9.10 (#30844)

fpruvost committed:
Pastix: new package (#30533)

adamjstewart committed:
py-scikit-learn: add v1.1.1 (#30830)

sethrj committed:
sed: fix recursive symlink (#30849)

Use `spack build` as build dir to avoid recursive link error.

```
config.status: linking /var/folders/fy/x2xtwh1n7fn0_0q2kk29xkv9vvmbqb/T/s3j/spack-stage/spack-stage-sed-4.8-wraqsot6ofzvr3vrgusx4mj4mya5xfux/spack-src/GNUmakefile to GNUmakefile
config.status: executing depfiles commands
config.status: executing po-directories commands
config.status: creating po/POTFILES
config.status: creating po/Makefile
==> sed: Executing phase: 'build'
==> [2022-05-25-14:15:51.310333] 'make' '-j8' 'V=1'
make: GNUmakefile: Too many levels of symbolic links
make: stat: GNUmakefile: Too many levels of symbolic links
make: *** No rule to make target `GNUmakefile'.  Stop.
```

snehring committed:
eagle: updating to version 1.1.3 (#30852)

dialvarezs committed:
Add nextflow 22.04.3 (#30855)

sethrj committed:
openmpi: fixes for slurm and #29449 (#30299)

adamjstewart committed:
py-pythran: add v0.11.0 (#30829)

adamjstewart committed:
py-numpy: add v1.22.4 (#30827)

tgamblin committed:
bugfix: do not compute `package_hash` for old concrete specs (#30861)

Old concrete specs were slipping through in `_assign_hash`, and `package_hash` was
attempting to recompute a package hash when we could not know the package a time
of concretization.

Part of this was that the logic for `_assign_hash` was hard to understand -- it was
called twice from `_finalize_concretization` and had special cases for both args it
was called with. It's much easier to understand the logic here if we just inline it.

- [x] Get rid of `_assign_hash` and just integrate it with `_finalize_concretization`
- [x] Don't call `_package_hash` at all for already-concrete specs.
- [x] Add regression test.

adamjstewart committed:
libtiff: remove extra dependencies/patch (#30854)

adamjstewart committed:
py-tensorboard: add v2.9.0 (#30832)

scottwittenburg committed:
ci: Support secure binary signing on protected pipelines (#30753)

This PR supports the creation of securely signed binaries built from spack
develop as well as release branches and tags. Specifically:

- remove internal pr mirror url generation logic in favor of buildcache destination
on command line
    - with a single mirror url specified in the spack.yaml, this makes it clearer where 
    binaries from various pipelines are pushed
- designate some tags as reserved: ['public', 'protected', 'notary']
    - these tags are stripped from all jobs by default and provisioned internally
    based on pipeline type
- update gitlab ci yaml to include pipelines on more protected branches than just
develop (so include releases and tags)
    - binaries from all protected pipelines are pushed into mirrors including the
    branch name so releases, tags, and develop binaries are kept separate
- update rebuild jobs running on protected pipelines to run on special runners
provisioned with an intermediate signing key
    - protected rebuild jobs no longer use "SPACK_SIGNING_KEY" env var to
    obtain signing key (in fact, final signing key is nowhere available to rebuild jobs)
    - these intermediate signatures are verified at the end of each pipeline by a new
    signing job to ensure binaries were produced by a protected pipeline
- optionallly schedule a signing/notary job at the end of the pipeline to sign all
packges in the mirror
    - add signing-job-attributes to gitlab-ci section of spack environment to allow
    configuration
    - signing job runs on special runner (separate from protected rebuild runners)
    provisioned with public intermediate key and secret signing key

alalazo committed:
Remove the warning that Spack prints at each spec (#30872)

Add instead a warning box in the documentation

drhansj committed:
Adding new package bricks for x86, cuda (#30863)

* Adding new package bricks for x86, cuda

* Fixed complaints from "spack style" that CI found

* add license comment at top

Co-authored-by: drhansj <drhansj@berkeley.edu>
Co-authored-by: eugeneswalker <38933153+eugeneswalker@users.noreply.github.com>

drkrynstrng committed:
Add newer nano versions (#30865)

drkrynstrng committed:
Add new versions of GNU parallel (#30862)

bvanessen committed:
Added hash for new versions (#30860)

simonpintarelli committed:
q-e-sirius: remove ~apps constraint (#30857)

rspavel committed:
Updated mfem constraints in laghos spackage (#30851)

Updated mfme constraints in laghos spackage to better match comments and
support legacy builds of `laghos@1.0:2.0`

scottwittenburg committed:
Revert "strip -Werror: all specific or none (#30284)" (#30878)

This reverts commit 330832c22cfa59554f6681a570bdec24ca46e79b.

severinstrobl committed:
otf2/scorep: add versions 3.0/7.1 (#28631)

prckent committed:
llvm: 14.0.3 and 14.0.4 (#30888)

adamjstewart committed:
Python: fix clingo bootstrapping on Apple M1 (#30834)

This PR fixes several issues I noticed while trying to get Spack working on Apple M1.

- [x] `build_environment.py` attempts to add `spec['foo'].libs` and `spec['foo'].headers` to our compiler wrappers for all dependencies using a try-except that ignores `NoLibrariesError` and `NoHeadersError` respectively. However, The `libs` and `headers` attributes of the Python package were erroneously using `RuntimeError` instead.
- [x] `spack external find python` (used during bootstrapping) currently has no way to determine whether or not an installation is `+shared`, so previously we would only search for static Python libs. However, most distributions including XCode/Conda/Intel ship shared Python libs. I updated `libs` to search for both shared and static (order based on variant) as a fallback.
- [x] The `headers` attribute was recursively searching in `prefix.include` for `pyconfig.h`, but this could lead to non-deterministic behavior if multiple versions of Python are installed and `pyconfig.h` files exist in multiple `<prefix>/include/pythonX.Y` locations. It's safer to search in `sysconfig.get_path('include')` instead.
- [x] The Python installation that comes with XCode is broken, and `sysconfig.get_paths` is hard-coded to return specific directories. This meant that our logic for `platlib`/`purelib`/`include` where we replace `platbase`/`base`/`installed_base` with `prefix` wasn't working and the `mkdirp` in `setup_dependent_package` was trying to create a directory in root, giving permissions issues. Even if you commented out those `mkdirp` calls, Spack would add the wrong directories to `PYTHONPATH`. Added a fallback hard-coded to `lib/pythonX.Y/site-packages` if sysconfig is broken (this is what distutils always did).

michaelkuhn committed:
gcc: add 9.5.0 (#30893)

drkrynstrng committed:
Add R 4.2.0 (#30859)

JDBetteridge committed:
Ensure same BLAS/LAPACK config from Numpy used in Scipy (#30818)

* Call Numpy package's set_blas_lapack() and setup_build_environment() in Scipy package

* Remove broken link from comment

* Use .package attribute of spec to avoid import

becker33 committed:
update tutorial command for v0.18.0 and new gpg key (#30904)

becker33 committed:
target optimization: re-norm optimization scale so that 0 is best. (#29926)

referred targets are currently the only minimization criteria for Spack for which we allow
negative values. That means Spack may be incentivized to add nodes to the DAG if they
match the preferred target.

This PR re-norms the minimization criteria so that preferred targets are weighted from 0,
and default target weights are offset by the number of preferred targets per-package to
calculate node_target_weight.

Also fixes a bug in the test for preferred targets that was making the test easier to pass
than it should be.

eschnett committed:
hpx: New version 1.8.0 (#30848)

JDBetteridge committed:
Additional BLAS/LAPACK library configuration for Numpy (#30817)

* Add amdblis and amdlibflame as BLAS/LAPACK options

* Add Cray-libsci as BLAS/LAPACK option

* Use Netlib config for Cray-libsci

tgamblin committed:
update changelog for v0.18.0 (#30905)

trws committed:
refactor: packages import `spack.package` explicitly (#30404)

Explicitly import package utilities in all packages, and corresponding fallout.

This includes:

* rename `spack.package` to `spack.package_base`
* rename `spack.pkgkit` to `spack.package`
* update all packages in builtin, builtin_mock and tutorials to include `from spack.package import *`
* update spack style
  * ensure packages include the import
  * automatically add the new import and remove any/all imports of `spack` and `spack.pkgkit`
    from packages when using `--fix`
  * add support for type-checking packages with mypy when SPACK_MYPY_CHECK_PACKAGES
    is set in the environment
* fix all type checking errors in packages in spack upstream
* update spack create to include the new imports
* update spack repo to inject the new import, injection persists to allow for a deprecation period

Original message below:
 
As requested @adamjstewart, update all packages to use pkgkit.  I ended up using isort to do this,
so repro is easy:

```console
$ isort -a 'from spack.pkgkit import *' --rm 'spack' ./var/spack/repos/builtin/packages/*/package.py
$ spack style --fix
```

There were several line spacing fixups caused either by space manipulation in isort or by packages
that haven't been touched since we added requirements, but there are no functional changes in here.

* [x] add config to isort to make sure this is maintained going forward

tgamblin committed:
0.18.0.dev0 -> 0.19.0.dev0 (#30907)

bollig committed:
Alinux isc buildcache (#30462)

Add two new stacks targeted at x86_64 and arm, representing an initial list of packages 
used by current and planned AWS Workshops, and built in conjunction with the ISC22
announcement of the spack public binary cache.

Co-authored-by: Scott Wittenburg <scott.wittenburg@kitware.com>

adamjstewart committed:
protobuf: fix spack versions (#30879)

jeanbez committed:
HDF5 VOL-ASYNC update versions (#30900)

bvanessen committed:
Add new versions of protobuf and py-protobuf (#30503)

* Add new versions

* Updated the hashes to match the published pypi.org hashes.  Added version constraints for Python.

joequant committed:
pass CC variable to make (#30912)

Set CC to cc

iarspider commented:
@tldahlgren thank you for your review. Some additional comments below:
> We're already generating a template for the package with, pre-PR, information derived from the archive file. We also support packages with archive urls and git.
> So why not add lines and guiding comments for the git assignment, etcetera?

Getting code from Git (or any other VCS) is, IMO, an advancaed topic and should be kept in the documentation as to not confuse package authors. This also reduces the number of TODO lines one has to remove before pushing the recipe upstream. 

> You could also use that to determine if the branch, commit, or tag options are valid IF you want.
I believe this will require checking out the code, which I have no idea how to do properly (my spack-foo is not that strong). This can be done in a follow-up PR, I think.

iarspider committed:
Redesign
